---
title: "R script for finding marker genes"
output:
  html_document: default
  html_notebook: default
---

Load packages

```{r}
suppressPackageStartupMessages({
   library(ggplot2)
})
```

Loading the data and normalising it
```{r}
testis.single.cell <- read.csv('data/Run2_91509_gene_count_matrix_Testis_ProteinCoding.csv', 
             row.names = 1)

bulk.data.raw <- list( 
   sl.rna = read.table('data/GSE43717_RAW/GSM1069639_mmu_sl_m.FPKM.txt', header = T, stringsAsFactors=F ),
   sc.rna = read.table('data/GSE43717_RAW/GSM1069641_mmu_sc_m.FPKM.txt', header = T, stringsAsFactors=F ),
   sg.rna = read.table('data/GSE43717_RAW/GSM1069640_mmu_sg_m.FPKM.txt', header = T, stringsAsFactors=F ),
   sd.rna = read.table('data/GSE43717_RAW/GSM1069642_mmu_sd_m.FPKM.txt', header = T, stringsAsFactors=F ),
   sz.rna = read.table('data/GSE43717_RAW/GSM1069643_mmu_sz_m.FPKM.txt', header = T, stringsAsFactors=F ) )
```

Some gene IDs are duplicated in the bulk data
```{r}
duplicated.genes <- Reduce( union, lapply( bulk.data.raw, function(l) l$gene_id[ duplicated(l$gene_id) ] ) )
```

Find common genes in all tables

```{r}
genes.common <- Reduce( intersect, lapply( bulk.data.raw, function(l) l$gene_id ), rownames(testis.single.cell) )
head( genes.common)
```

Construct FPKM table

```{r}
testis.types.rna <- sapply( bulk.data.raw, function( l ){
   ind <- match( genes.common, l$gene_id )
   l$FPKM[ind]
})
rownames( testis.types.rna ) <- genes.common
colnames(testis.types.rna) <- sub( ".rna", "", colnames(testis.types.rna) )
head(testis.types.rna)
```

Subset single-cell data to common genes

```{r}
testis.single.cell <- as.matrix( testis.single.cell[genes.common, ] )
```

Normalize using GMPR by Chen and Chen, 2017 (https://doi.org/10.1101/112565):

```{r}
source('GMPR.R')
size.gmpr <- GMPR(testis.single.cell)
testis.single.cell.gmpr <- t(t(testis.single.cell)/size.gmpr$gmpr)
```

Plot size factors

```{r}
plot( size.gmpr$gmpr, log="y" )
```

```{r}
testis.types.rna.log <- log2( testis.types.rna + 1 )
```

Next we write the function for calculating the spread and difference for each gene for
each cell type in RNA Seq data. The spread for each gene for each cell type is calculated by computing the difference between the maximum and minimum of log expression values of cell types other than the type of interest. The difference is calculated by computing the difference between log expression values of the cell type of interest and mean of log expression values for all other cell types. Above is done for each gene for each cell.

```{r}
get.spread.diff <- function(testis.types.rna.log) #testis.types.rna.log = the data frame containing the 5 types of rna
{
  testis.type.spread.diff <- sapply( colnames(testis.types.rna.log),
                                     function(toi) {   # toi = type of interest
   others <- testis.types.rna.log[ , colnames(testis.types.rna.log) != toi ]
   data.frame( 
      spread = apply( others, 1, function(r) max(r) - min(r) ),
      diff =  testis.types.rna.log[,toi] - rowMeans( others ) ) }, simplify=FALSE )
  return(testis.type.spread.diff)  
}
testis.type.spread.diff <- get.spread.diff(testis.types.rna.log)
str(testis.type.spread.diff)
```


Plot for one type
```{r}
plot( testis.type.spread.diff$sz$spread, testis.type.spread.diff$sz$diff, col = '#00000050' )
abline(0,2)
abline(0,-2)
abline(h = c(-2,2))
```



For each cell type the points of interest are points above the respective lines and having absolute slopes greater than resepective threshold. For our case we have kept it at 2. get.points.beyond computes the genes following the criteria
and genes.of.int are those genes for each cell type.
```{r}
get.points.beyond <- function(testis.type, slope)  # testis.type is one element of the list 'testis.types.spread.diff
{
  return(which(abs(testis.type$diff/testis.type$spread)  > slope & abs(testis.type$diff) > 2))
}

genes.of.int <- sapply(testis.type.spread.diff, 
   function(testis.type) {
     get.points.beyond(testis.type, 2)
   },
   simplify=FALSE )

str(genes.of.int)
```


Now once we have found the marker genes we try to label the single cell sequencing data. For that we first subtract the 
means from the respective samples for both single cell and rna seq for each gene.

```{r}
testis.types.rna.log.centered <- testis.types.rna.log - rowMeans( testis.types.rna.log )
testis.single.cell.log.centered <- log2( testis.single.cell.gmpr + 1 ) - rowMeans( log2( testis.single.cell.gmpr + 1 ) )

centered.data <- list()
centered.data[['bulk']] <- testis.types.rna.log.centered
centered.data[['single.cell']] <- testis.single.cell.log.centered
```

Marker Genes
```{r}
all.marker.genes <- unname( unlist( genes.of.int ) )
```

Getting the covariance
```{r}
get.covariance <- function(single.cell.data, bulk.data, genes)
{
  return(t(single.cell.data[genes, ]) %*% bulk.data[genes,])
}
covariances <- get.covariance(testis.single.cell.log.centered, testis.types.rna.log.centered, all.marker.genes)
head( covariances)
```

Get labels

```{r}
get.labels <- function(cov.mat)
{
  labels <- colnames(covariances)[ apply( covariances, 1, which.max ) ]
  names(labels) <- rownames(cov.mat)  
  return(labels)
}
labels <- get.labels(covariances)
head(labels)
table( labels )
```

Plot largest to second largest

```{r}
get.good.labels <- function(cov.mat, labels, threshold, plot = T)
{
  df <- data.frame( 
   max = apply( cov.mat, 1, max ),
   second = apply( cov.mat, 1, function(x) sort(x,decreasing=T)[2] ),
   label = labels )

if(plot)
{
  ggplot( df , aes( x=max, y=second, col=label )) +
  geom_point() +
  geom_abline( intercept = threshold )
}
return(df$max - df$second > -threshold )
}


```

Which labels are good (to the right of the line)

```{r}
label_good <- get.good.labels(covariances, labels, -50)
```

Another way of getting the labels: Use only a type's own markers to calculate covariance:

```{r}
toi <- "sz"

# using all marker genes, as above
covariances1b <-
sapply( colnames(testis.types.rna.log), function(toi)
   testis.types.rna.log.centered[ all.marker.genes, toi ] %*% testis.single.cell.log.centered[ all.marker.genes, ] )
stopifnot( all( abs( covariances1b - covariances ) < 1e-10 ) )  # still the same?

# the same, using onlye this type's markerss
covariances2 <-
sapply( colnames(testis.types.rna.log), function(toi)
   testis.types.rna.log.centered[ genes.of.int[[toi]], toi ] %*% testis.single.cell.log.centered[ genes.of.int[[toi]], ] )

# get new labels
labels2 <- colnames(covariances2)[ apply( covariances2, 1, which.max ) ]
table( labels, labels2, label_good )
```

WORKED TILL HERE


```{r}
labels.single.cell <- read.table(file = 'data/labels_testis_single_cell.txt')
labels.single.cell$rna <- sapply(as.character(labels.single.cell$V3), function(x)
  {
  if(x == levels(labels.single.cell$V3)[1])
    'sd'
  else if(x == levels(labels.single.cell$V3)[2])
    'sd'
  else if(x == levels(labels.single.cell$V3)[3])
    'sl'
  else if(x == levels(labels.single.cell$V3)[4])
    'sc'
  else if(x == levels(labels.single.cell$V3)[5])
    'sg'
  else
    'sz'
})
print("Current Mapping of our labels with Florent's data")
sum(labels.single.cell$rna == labels)

table(ours = labels, florent = labels.single.cell$rna)
```

To be more sure about our labels we look at the second maximum value of df and compare it with the maximum.
```{r}
max.cells <- apply(df, 1, max)
sec.max.cells <- apply(df, 1, function(x)
  {
    x[order(x, decreasing = T)[2]]
})
{plot(max.cells, sec.max.cells)
abline(0,1)
abline(-50,1)}

```

We look at the above graph and decide a threshold of -50 such that all points below that are samples whose label we can be sure of. Below function gives us the labels based on this.

```{r}
get.sure.samples <- function(df, threshold, labels.existing)
{
  max.cells <- apply(df, 1, max)
  sec.max.cells <- apply(df, 1, function(x)
  {
    x[order(x, decreasing = T)[2]]
  })
  labs.mod <- sapply(seq_along(labels.existing), function(i)
  {
    val = sec.max.cells[i] - max.cells[i] - threshold
    if(val >= 0)
      'other'
    else
      labels.existing[i]
  }) 
}

labels.mod <- get.sure.samples(df, -50, labels)
table(labels.mod)
```

Just to be sure about this method we also randomly divide the genes into sets of two and rerun the above analysis.

```{r}
get.output <- function(centered.data, marker.genes, genes.common)
{
  pair1 <- sample(genes.common, length(genes.common)/2)
  pair2 <- setdiff(genes.common, pair1)

  marker.genes.list <- lapply(list(pair1, pair2), function(pair)
    {
    intersect(genes.common[marker.genes], pair)
  })

  dfs <- lapply(marker.genes.list, function(marker.genes) {
    get.covariance(centered.data$single.cell, centered.data$bulk, marker.genes)})
        
  labs <- lapply(dfs, get.labels)
  labs.mod <- lapply(seq_along(labs), function(i) get.good.labels(dfs[[i]], labs[[i]], -50))

  return(list(labels = labs, labels.mod = labs.mod, marker.genes = marker.genes.list,
              mean.data = req.mean.data))
}

```

We first compute the mean for the entire gene set
```{r}
mean.entire.data <- get.means(log2(testis.single.cell.gmpr+1), 
                              testis.types.rna.log.df, genes.common)
                   
```

We then find labels for a random sampling of genes and then try to evaluate the
labels we get
```{r}
out1 =  get.output(centered.data, unlist(genes.of.int), genes.common)
table(out1$labels[[1]])
table(out1$labels[[2]])

sum(out1$labels[[1]] == out1$labels[[2]])
sum(out1$labels.mod[[1]] == out1$labels.mod[[2]])
sum(out1$labels.mod[[1]] == out1$labels[[2]])

table(out1$labels.mod[[1]])
table(out1$labels.mod[[2]])

```
